File: nodejs_implementation/src/scraper/direct-markdown-scraper.js
/**
 * Direct Markdown Scraper for DeepWiki to Markdown Converter
 * 
 * This module provides functionality for directly scraping Markdown content from DeepWiki sites.
 */

const axios = require('axios');
const { URL } = require('url');
const cheerio = require('cheerio');
const { BaseScraper } = require('./base-scraper');
const { Content } = require('../models/content');
const { logger } = require('../utils/logger');
const { handleRequestError, retryOnException } = require('../utils/error-handler');
const { fixMarkdownLinks } = require('../utils/link-processor');

/**
 * Scraper for directly extracting Markdown content from DeepWiki sites.
 * 
 * This scraper assumes that the DeepWiki site provides Markdown content directly,
 * without the need for HTML-to-Markdown conversion.
 */
class DirectMarkdownScraper extends BaseScraper {
  /**
   * Initialize the DirectMarkdownScraper.
   * 
   * @param {Object} config - Configuration object containing scraper settings
   */
  constructor(config) {
    super(config);
    this.session = axios.create({
      headers: config.headers,
      timeout: config.timeout
    });
    this.visitedUrls = new Set();
  }
  
  /**
   * Scrape Markdown content directly from the DeepWiki site.
   * 
   * @returns {Promise<Array<Object>>} List of scraped content items
   */
  async scrape() {
    const results = [];
    
    // Start with the main URL
    const mainUrl = this.config.url;
    if (!this._validateUrl(mainUrl)) {
      logger.error(`Invalid URL: ${mainUrl}`);
      return results;
    }
    
    // Process the main page
    logger.info(`Scraping content from ${mainUrl}`);
    const content = await this._fetchContent(mainUrl);
    
    if (content) {
      // Process the main content
      const result = {
        url: mainUrl,
        title: this._extractTitle(content),
        content: this._cleanContent(content),
        library: this.config.libraryName
      };
      results.push(result);
      this.visitedUrls.add(mainUrl);
      
      // Extract and process navigation links
      const navItems = this.extractNavigation(content);
      for (const navItem of navItems) {
        if (!this.visitedUrls.has(navItem.url)) {
          // Recursively process navigation items
          const navResults = await this._processNavigationItem(navItem);
          results.push(...navResults);
        }
      }
    }
    
    return results;
  }
  
  /**
   * Process a navigation item by scraping its content.
   * 
   * @param {Object} navItem - Navigation item with title and url
   * @returns {Promise<Array<Object>>} List of scraped content items
   * @private
   */
  async _processNavigationItem(navItem) {
    const results = [];
    const url = navItem.url;
    
    if (this.visitedUrls.has(url)) {
      return results;
    }
    
    logger.debug(`Processing navigation item: ${navItem.title} (${url})`);
    
    // Fetch content for the navigation item
    const content = await this._fetchContent(url);
    if (content) {
      // Process the content
      const result = {
        url: url,
        title: navItem.title || this._extractTitle(content),
        content: this._cleanContent(content),
        library: this.config.libraryName
      };
      results.push(result);
      this.visitedUrls.add(url);
      
      // Extract and process nested navigation links
      const nestedNavItems = this.extractNavigation(content);
      for (const nestedItem of nestedNavItems) {
        if (!this.visitedUrls.has(nestedItem.url)) {
          const nestedResults = await this._processNavigationItem(nestedItem);
          results.push(...nestedResults);
        }
      }
    }
    
    return results;
  }
  
  /**
   * Fetch content from the specified URL.
   * 
   * @param {string} url - URL to fetch content from
   * @returns {Promise<string|null>} Fetched content or null if an error occurred
   * @private
   */
  async _fetchContent(url) {
    try {
      const response = await this._fetchWithRetry(url);
      return response.data;
    } catch (error) {
      handleRequestError(error, url);
      return null;
    }
  }
  
  /**
   * Fetch content with retry logic.
   * 
   * @param {string} url - URL to fetch content from
   * @returns {Promise<Object>} Axios response object
   * @private
   */
  _fetchWithRetry = retryOnException(
    async (url) => {
      return await this.session.get(url);
    },
    {
      maxRetries: this.config.retryCount,
      retryDelay: this.config.retryDelay,
      retryOn: [axios.AxiosError]
    }
  );
  
  /**
   * Extract title from the content.
   * 
   * @param {string} content - Content to extract title from
   * @returns {string} Extracted title or default title
   * @private
   */
  _extractTitle(content) {
    // Try to extract title from the first heading
    const match = content.match(/^#\s+(.+)$/m);
    if (match) {
      return match[1].trim();
    }
    
    // Fallback to a default title
    return 'Untitled Document';
  }
  
  /**
   * Extract navigation links from Markdown content.
   * 
   * @param {string} content - Markdown content to extract navigation from
   * @returns {Array<Object>} List of navigation items with title and url
   */
  extractNavigation(content) {
    const navItems = [];
    
    // Extract Markdown links
    const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
    let match;
    
    while ((match = linkPattern.exec(content)) !== null) {
      const title = match[1].trim();
      const url = match[2].trim();
      
      // Skip empty or anchor links
      if (!url || url.startsWith('#')) {
        continue;
      }
      
      // Handle relative URLs
      let fullUrl = url;
      if (!url.startsWith('http://') && !url.startsWith('https://')) {
        fullUrl = new URL(url, this.config.url).href;
      }
      
      navItems.push({
        title: title,
        url: fullUrl
      });
    }
    
    return navItems;
  }
  
  /**
   * Clean the scraped content.
   * 
   * @param {string} content - Content to clean
   * @returns {string} Cleaned content
   */
  _cleanContent(content) {
    // Basic cleaning
    let cleaned = super._cleanContent(content);
    
    // Fix Markdown links
    cleaned = fixMarkdownLinks(cleaned);
    
    return cleaned;
  }
  
  /**
   * Create Content objects from scraped data.
   * 
   * @param {Array<Object>} scrapedItems - List of scraped items
   * @returns {Array<Content>} List of Content objects
   */
  createContentObjects(scrapedItems) {
    return scrapedItems.map(item => new Content({
      title: item.title,
      content: item.content,
      url: item.url,
      library: item.library,
      filePath: this._createFilePath(item)
    }));
  }
}

module.exports = {
  DirectMarkdownScraper
};

File: deepwiki_to_md/scraper/direct_markdown_scraper.py
"""
DeepWiki to Markdownコンバーター用のダイレクトMarkdownスクレイパー。

このスクレイパーはDeepWikiサイトから直接Markdownコンテンツを取得します。
"""

import logging
import os
import re
import time
from typing import List, Dict, Any, Optional
from urllib.parse import urljoin, urlparse

import requests

from deepwiki_to_md.models.config import Config
from deepwiki_to_md.scraper.base_scraper import BaseScraper

logger = logging.getLogger(__name__)


class DirectMarkdownScraper(BaseScraper):
    """
    DeepWikiサイトから直接Markdownコンテンツを取得するためのスクレイパー。
    
    このスクレイパーは、DeepWikiサイトがHTMLの解析や変換を必要とせずに、
    Markdownコンテンツを直接提供することを前提としています。
    """

    def __init__(self, config: Config):
        """
        指定された設定でスクレイパーを初期化します。
        
        Args:
            config (Config): URL、ライブラリ名、出力ディレクトリを含む設定オブジェクト。
        """
        super().__init__(config)
        self.session = requests.Session()
        self.session.headers.update(config.headers)

    def scrape(self) -> List[Dict[str, Any]]:
        """
        DeepWikiサイトからコンテンツをスクレイピングします。
        
        Returns:
            List[Dict[str, Any]]: スクレイピングされたコンテンツを含む辞書のリスト。
        """
        results = []

        # メインコンテンツを取得
        content = self._fetch_content(self.config.url)
        if not content:
            logger.error(f"Failed to fetch content from {self.config.url}")
            return results

        # メインコンテンツを保存
        filename = self._get_filename_from_url(self.config.url)
        filepath = self.save_content(filename, content)
        results.append({
            "url": self.config.url,
            "filename": filename,
            "filepath": filepath
        })

        # ナビゲーション項目を抽出して処理
        nav_items = self.extract_navigation(content)
        for item in nav_items:
            if len(results) >= self.config.max_depth:
                logger.info(f"Reached maximum depth of {self.config.max_depth}")
                break

            url = item.get("url")
            if not url:
                continue

            # 処理済みの場合はスキップ
            if any(r["url"] == url for r in results):
                continue

            # コンテンツを取得して保存
            item_content = self._fetch_content(url)
            if not item_content:
                logger.warning(f"Failed to fetch content from {url}")
                continue

            item_filename = item.get("title", self._get_filename_from_url(url))
            item_filepath = self.save_content(item_filename, item_content)
            results.append({
                "url": url,
                "filename": item_filename,
                "filepath": item_filepath
            })

            # リクエスト間の遅延を考慮
            time.sleep(self.config.delay)

        return results

    def extract_navigation(self, content: str) -> List[Dict[str, Any]]:
        """
        コンテンツからナビゲーションを抽出します。
        
        Args:
            content (str): ナビゲーションを抽出するコンテンツ。
            
        Returns:
            List[Dict[str, Any]]: ナビゲーション項目を含む辞書のリスト。
        """
        nav_items = []

        # Markdownコンテンツからリンクを抽出
        # これはMarkdownからリンクを抽出する単純な実装です
        # より堅牢な実装ではMarkdownパーサーを使用します
        link_pattern = r'\[([^\]]+)\]\(([^)]+)\)'
        for match in re.finditer(link_pattern, content):
            title = match.group(1)
            url = match.group(2)

            # DeepWiki以外のリンクはスキップ
            if "deepwiki.com" not in url and not url.startswith("/"):
                continue

            # 相対URLを絶対URLに変換
            if url.startswith("/"):
                url = urljoin(self.config.url, url)

            nav_items.append({
                "title": title,
                "url": url
            })

        return nav_items

    def _fetch_content(self, url: str) -> Optional[str]:
        """
        指定されたURLからコンテンツを取得します。
        
        Args:
            url (str): コンテンツを取得するURL。
            
        Returns:
            Optional[str]: 取得したコンテンツ。リクエストが失敗した場合はNone。
        """
        retries = 0
        while retries <= self.config.retry_limit:
            try:
                if self.config.verbose:
                    logger.info(f"Fetching content from {url}")

                response = self.session.get(url, timeout=self.config.timeout)
                response.raise_for_status()

                return response.text
            except requests.RequestException as e:
                retries += 1
                if retries > self.config.retry_limit:
                    logger.error(f"Failed to fetch content from {url}: {e}")
                    return None

                logger.warning(f"Retry {retries}/{self.config.retry_limit} for {url}: {e}")
                time.sleep(self.config.delay * (2 ** (retries - 1)))  # 指数関数的バックオフ

        return None

    def _get_filename_from_url(self, url: str) -> str:
        """
        指定されたURLからファイル名を取得します。
        
        Args:
            url (str): ファイル名を取得するURL。
            
        Returns:
            str: URLから派生したファイル名。
        """
        parsed = urlparse(url)
        path = parsed.path.strip("/")

        if not path:
            return "index"

        # パスの最後の部分をファイル名として使用
        filename = path.split("/")[-1]

        # ファイル拡張子が存在する場合は削除
        filename = os.path.splitext(filename)[0]

        return filename
